### PR TITLE
Close #7059: Implement landscape doors on Ghost Train

### LINKED ISSUE #7059: Doors in land textures for ghost train rides do not appear when loading SV4
In vanilla RCTLL, doors appear in the land textures when a ghost train ride passes through. In OpenRCT2, it appears that this behavior is not carried over, as the areas that are supposed to be doors are holes. See the comparison image below. I'm not sure if this can be fixed given the limitations of the SV6 file format, but I thought I would report it nonetheless.

RCTLL:
capture

OpenRCT2:
aceworkbench_v11 2018-01-19 18-02-30

### PR DESCRIPTION
**Pull request by Gymnasiast**
afbeelding

### CODE REVIEW

File: src/openrct2/ride/Vehicle.cpp [hunk 1]
```diff
@@ -7487,6 +7487,43 @@ void Vehicle::UpdateSceneryDoor() const
...
+            trackElement->SetDoorAState(LANDSCAPE_DOOR_CLOSED);
+        else
+            trackElement->SetDoorBState(LANDSCAPE_DOOR_CLOSED);
+        // TODO: play door close sound
+    }
+}
+
+void Vehicle::UpdateLandscapeDoor() const
+{
+    if (!GetRide()->GetRideTypeDescriptor().HasFlag(RIDE_TYPE_FLAG_HAS_LANDSCAPE_DOORS))
```
Comment (IntelOrca) [inline src/openrct2/ride/Vehicle.cpp:7514]: Be as defensive as possible, check for nullptr after calling `GetRide`.

File: src/openrct2/ride/gentle/GhostTrain.cpp [hunk 2]
```diff
@@ -127,6 +127,42 @@ static constexpr const uint32_t ghost_train_track_pieces_brakes[4] = {
     SPR_GHOST_TRAIN_TRACK_BRAKES_NW_SE,
 };
 
+static uint8_t doorOpeningOutwardsToImage[] = {
```
Comment (IntelOrca) [inline src/openrct2/ride/gentle/GhostTrain.cpp:130]:
```suggestion
static constexpr const uint8_t doorOpeningOutwardsToImage[] = {
```

File: src/openrct2/ride/gentle/GhostTrain.cpp [hunk 3]
```diff
@@ -127,6 +127,42 @@ static constexpr const uint32_t ghost_train_track_pieces_brakes[4] = {
...
+    TUNNEL_DOORS_2, // Closed
+    TUNNEL_DOORS_2, // Unused?
+    TUNNEL_DOORS_3, // Half open
+    TUNNEL_DOORS_4, // Fully open
+    TUNNEL_DOORS_2, // Unused?
+    TUNNEL_DOORS_2, // Unused?
+    TUNNEL_DOORS_2, // Unused?
+};
+
+static uint8_t doorOpeningInwardsToImage[] = {
```
Comment (IntelOrca) [inline src/openrct2/ride/gentle/GhostTrain.cpp:140]:
```suggestion
static constexpr const uint8_t doorOpeningInwardsToImage[] = {
```

File: src/openrct2/ride/gentle/GhostTrain.cpp [hunk 4]
```diff
@@ -329,6 +381,8 @@ static void paint_ghost_train_station(
 
     paint_util_push_tunnel_rotated(session, direction, height, TUNNEL_SQUARE_FLAT);
 
+    //    const uint8_t tunnelDoors = get_tunnel_doors_image(tileElement, direction);
+
```
Comment (IntelOrca) [inline src/openrct2/ride/gentle/GhostTrain.cpp:385]: Commented out code

File: src/openrct2/rct12/RCT12.h [hunk 5]
```diff
@@ -511,10 +511,11 @@ struct RCT12TrackElement : RCT12TileElementBase
     uint8_t GetSeatRotation() const;
     uint16_t GetMazeEntry() const;
     uint8_t GetPhotoTimeout() const;
-    // Used in RCT1, will be reintroduced at some point.
-    // (See https://github.com/OpenRCT2/OpenRCT2/issues/7059)
+    // RCT1 feature, reintroduced by OpenRCT2. See https://github.com/OpenRCT2/OpenRCT2/issues/7059)
```
Comment (Gymnasiast) [inline src/openrct2/rct12/RCT12.h:514]:
```suggestion
    // RCT1 feature, reintroduced by OpenRCT2. See https://github.com/OpenRCT2/OpenRCT2/issues/7059
```

**Comment by duncanspumpkin** [comment]
Doesn't this change all existing ghost rides i.e. sv6 and sv4?

**Comment by Gymnasiast** [comment]
Yes, that's the whole point: the stucco and sandstone terrain edges (which were not in RCT2) should have doors for the Ghost Train.

There should not be any visual difference for regular terrain edges (like rock, ice, brown poles, black poles).

**Comment by duncanspumpkin** [comment]
ahh okay that sounds fine then.

**Review by duncanspumpkin — APPROVED** [review]

**Pull request merged**